java-frontend/src/main/java/org/sonar/java/se/checks/CustomUnclosedResourcesCheck.java

```diff
@@ -0,0 +1,453 @@
+
+  private enum Status {
+    OPENED, CLOSED
+  }
+
+  @RuleProperty(
+    key = "constructor",
+    description = "List of classes (full qualified name) used to create instances that need to be closed." + ARGUMENT_DESCRIPTION)
+  public String constructor = "";
+
+  @RuleProperty(
+    key = "factoryMethod",
+    description = "List of classes (full qualified name#methodName) used to open instances that need to be closed." + ARGUMENT_DESCRIPTION)
+  public String factoryMethod = "";
+
+  @RuleProperty(
+    key = "openingMethod",
+    description = "List of methods (full qualified name#methodName) used to create instances that need to be closed." + ARGUMENT_DESCRIPTION)
+  public String openingMethod = "";
+
+  @RuleProperty(
+    key = "closingMethod",
+    description = "List of methods (full qualified name#methodName) used to close instances that need to be closed." + ARGUMENT_DESCRIPTION)
+  public String closingMethod = "";
+
+  private List<ConstructorMatcher> classConstructor;
+  private List<MethodMatcher> factoryList;
+  private List<MethodMatcher> openingList;
+  private List<MethodMatcher> closingList;
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    final PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    final PostStatementVisitor visitor = new PostStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  @Override
+  public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager constraintManager) {
+    final List<ObjectConstraint> constraints = context.getState().getFieldConstraints(Status.OPENED);
+    for (ObjectConstraint constraint : constraints) {
+      Tree syntaxNode = constraint.syntaxNode();
+      String name = null;
+      if (syntaxNode.is(Tree.Kind.NEW_CLASS)) {
+        name = ((NewClassTree) syntaxNode).identifier().symbolType().name();
+      } else if (syntaxNode.is(Tree.Kind.METHOD_INVOCATION)) {
+        name = ((MethodInvocationTree) syntaxNode).symbolType().name();
+      }
+      if (name != null) {
+        context.reportIssue(syntaxNode, this, "Close this \"" + name + "\".");
+      }
+    }
+  }
+
+  boolean isOpeningResource(MethodInvocationTree syntaxNode) {
+    for (MethodMatcher matcher : openingMethods()) {
+      if (matcher.matches(syntaxNode)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  List<ConstructorMatcher> constructorClasses() {
+    if (classConstructor == null) {
+      if (constructor.length() > 0) {
+        classConstructor = ImmutableList.of(new ConstructorMatcher(constructor));
+      } else {
+        classConstructor = ImmutableList.of();
+      }
+    }
+    return classConstructor;
+  }
+
+  List<MethodMatcher> factoryMethods() {
```

Comment:
The 3 methods here share the EXACT same code except for the parameter. Wild guess but I think some refactoring is in order (might also be factorized with constructorClasses method).